include/zephyr/drivers/sensor_v2.h

```diff
@@ -0,0 +1,878 @@
+ * @param buffer The buffer to use
+ * @return 0 on success
+ * @return < 0 on failure
+ */
+__syscall int sensor_set_fifo_data_buffer(const struct device *sensor, sensor_fifo_data_t *buffer);
+
+static inline int z_impl_sensor_set_fifo_data_buffer(const struct device *sensor,
+						     sensor_fifo_data_t *buffer)
+{
+#ifdef CONFIG_SENSOR_STREAMING_MODE
+	const struct sensor_driver_api_v2 *api = sensor->api;
+
+	if (api == NULL || api->set_fifo_data_buffer == NULL) {
+		return -ENOSYS;
+	}
+
+	return api->set_fifo_data_buffer(sensor, buffer);
+#else
+	return -ENOSYS;
+#endif
+}
+/**
+ * @brief Set the processing function for new data.
+ *
+ * Set the sensor's function for processing data. This callback is important
+ * because sensors which make use of a FIFO should not also need to cache the data
+ * in memory while waiting. Instead, these sensors will rely on this callback to
+ * notify the application of the new data. If the application did not set the
+ * callback, the data will be discarded.
+ *
+ * @param sensor Pointer to the sensor device
+ * @param callback A callback function to use when new data is available
+ * @return 0 on success
+ * @return < 0 on failure
+ */
+__syscall int sensor_set_process_data_callback(const struct device *sensor,
+					       sensor_process_data_callback_t callback);
+
+static inline int z_impl_sensor_set_process_data_callback(const struct device *sensor,
+							  sensor_process_data_callback_t callback)
+{
+#ifdef CONFIG_SENSOR_STREAMING_MODE
+	const struct sensor_driver_api_v2 *api = sensor->api;
+
+	if (api == NULL || api->set_process_data_callback == NULL) {
+		return -ENOSYS;
+	}
+
+	return api->set_process_data_callback(sensor, callback);
+#else
+	return -ENOSYS;
+#endif
+}
+
+__syscall int sensor_raw_reading_to_sensor_data(const struct device *sensor,
+						const sensor_fifo_data_t *reading,
+						uint16_t readings_offset, struct sensor_data *out);
+
+static inline int z_impl_sensor_raw_reading_to_sensor_data(const struct device *sensor,
+							   const sensor_fifo_data_t *reading,
+							   uint16_t readings_offset,
+							   struct sensor_data *out)
+{
+#ifdef CONFIG_SENSOR_STREAMING_MODE_RAW
+	const struct sensor_driver_api_v2 *api = sensor->api;
+
+	if (api == NULL || api->raw_reading_to_sensor_data == NULL) {
+		return -NOSYS;
+	}
+
+	return api->raw_reading_to_sensor_data(sensor, reading, reading_offset, out);
+#else
+	/* If there's no raw streaming mode, then 'reading' and 'out' are actually the same type. */
+	const struct sensor_data *in = reading;
+	int count = MIN(in->header.reading_count - readings_offset, out->header.reading_size);
+
+	if (count <= 0) {
+		return -EINVAL;
+	}
+
+	out->header.base_timestamp = in->header.base_timestamp;
+	out->header.reading_count = count;
+	memcpy(out->readings, in->readings + readings_offset,
+	       count * sizeof(struct sensor_sample_data));
+
+	return 0;
+#endif
+}
+
+/**
+ * @brief Get the available sample rates for the sensor
```

Comment:
This can become complicated: 
* some sensors support different power mode. It may be beneficial to know the power usage for a given sampling frequency and decide the most efficient one for the given application.
* some sensors support a fixed list of frequencies, other may be variable (when the driver is polling the sensor at regular interval). 

Given it is constant and given for a particular sensor, Could this be defined in a device tree? We would define helper functions to retrieve the best frequency to use for a given need.

Reply:
This is also true about trying to set the sample rate without setting other parameters (power mode, channels enabled), it leads to complexities.